File: tools/test/pricing.test.ts
import { describe, it, expect } from "vitest";

import { dataset } from "../src/data/index.js";
import { baseUnitPoints, pointsTierMissing } from "../src/data/pricing.js";

// World Eaters Chaos Terminators are priced by army ordinal: 175 for your 1st–2nd
// copy, 185 for your 3rd+ at 5 models, and 350/360 for a 6–10 model squad (a
// range tier). The id is shared with Emperor's Children, so resolve the WE copy.
const ct = dataset.units.getInFaction("chaos-terminators", "world-eaters")!.raw;

describe("baseUnitPoints — ordinal bands", () => {
  it("prices the 1st–2nd army copy at the lower band", () => {
    expect(baseUnitPoints(ct, 5, 1)).toBe(175);
    expect(baseUnitPoints(ct, 5, 2)).toBe(175);
    expect(baseUnitPoints(ct, 10, 1)).toBe(350);
  });

  it("prices the 3rd+ army copy at the higher band", () => {
    expect(baseUnitPoints(ct, 5, 3)).toBe(185);
    expect(baseUnitPoints(ct, 10, 3)).toBe(360);
    expect(baseUnitPoints(ct, 5, 7)).toBe(185); // open-ended top band
  });

  it("defaults to the 1st army copy when no ordinal is given", () => {
    expect(baseUnitPoints(ct, 5)).toBe(175);
  });

  it("picks the highest model tier the count reaches, within the band", () => {
    expect(baseUnitPoints(ct, 10, 1)).toBe(350);
    expect(baseUnitPoints(ct, 7, 1)).toBe(350); // inside the 6–10 range tier
    expect(baseUnitPoints(ct, 4, 1)).toBe(175); // below smallest tier → lowest tier
  });

  it("ignores ordinal for an unbanded unit (no unit_count_min)", () => {
    const bz = dataset.units.get("khorne-berzerkers")!.raw;
    expect(baseUnitPoints(bz, 10, 1)).toBe(baseUnitPoints(bz, 10, 99));
  });
});

describe("pointsTierMissing — ordinal-aware", () => {
  it("is false for a covered model count + ordinal, true below the smallest tier", () => {
    expect(pointsTierMissing(ct, 5, 1)).toBe(false);
    expect(pointsTierMissing(ct, 5, 3)).toBe(false);
    expect(pointsTierMissing(ct, 4, 1)).toBe(true);
  });
});

// Venatari Custodians are a GW range-priced unit: 3 models @160, or 4–6 models
// @320 (block pricing). The 320 tier carries models_max=6, so every size in its
// range prices at 320 — the regression that crashed the list builder on resize
// was 4/5-model squads falling through to the 3-model tier (160).
const ven = dataset.units.getInFaction("venatari-custodians", "adeptus-custodes")!.raw;

describe("range-priced tiers (models_max)", () => {
  it("prices every size in a range tier at that tier's cost", () => {
    expect(ven.points).toEqual([
      { models: 3, cost: 160 },
      { models: 4, models_max: 6, cost: 320 },
    ]);
    expect(baseUnitPoints(ven, 3)).toBe(160);
    expect(baseUnitPoints(ven, 4)).toBe(320);
    expect(baseUnitPoints(ven, 5)).toBe(320);
    expect(baseUnitPoints(ven, 6)).toBe(320);
  });

  it("flags counts outside every tier range (below floor, above ceiling)", () => {
    expect(pointsTierMissing(ven, 2)).toBe(true); // below the 3-model tier
    expect(pointsTierMissing(ven, 3)).toBe(false);
    expect(pointsTierMissing(ven, 4)).toBe(false);
    expect(pointsTierMissing(ven, 6)).toBe(false);
    expect(pointsTierMissing(ven, 7)).toBe(true); // above the 6-model ceiling
  });
});

File: go/pricing_test.go
package wh40kdc

import "testing"

// World Eaters Chaos Terminators are priced by army ordinal: 175 for the 1st-2nd
// copy, 185 for the 3rd+ (350/360 at 10 models). The id is shared with Emperor's
// Children, so resolve the WE copy. Mirror of the TS/Rust/Python pricing tests.
func weChaosTerminators(t *testing.T) map[string]any {
	t.Helper()
	ds := EmbeddedDataset()
	ct, ok := ds.Units.GetInFaction("chaos-terminators", "world-eaters")
	if !ok {
		t.Fatal("WE chaos-terminators not in dataset")
	}
	return ct.Raw
}

func TestBaseUnitPointsOrdinalBands(t *testing.T) {
	ct := weChaosTerminators(t)
	cases := []struct {
		models, ordinal, want int
	}{
		{5, 1, 175}, {5, 2, 175}, {10, 1, 350},
		{5, 3, 185}, {10, 3, 360}, {5, 7, 185},
		{7, 1, 350}, // inside the 6-10 range tier (not the 5-model tier)
	}
	for _, c := range cases {
		if got := baseUnitPoints(ct, c.models, c.ordinal); got != c.want {
			t.Errorf("baseUnitPoints(%d models, ordinal %d) = %d, want %d", c.models, c.ordinal, got, c.want)
		}
	}
}

func TestBaseUnitPointsUnbandedIgnoresOrdinal(t *testing.T) {
	ds := EmbeddedDataset()
	bz, ok := ds.Units.Get("khorne-berzerkers")
	if !ok {
		t.Fatal("khorne-berzerkers not in dataset")
	}
	if a, b := baseUnitPoints(bz.Raw, 10, 1), baseUnitPoints(bz.Raw, 10, 99); a != b {
		t.Errorf("unbanded unit should ignore ordinal: %d != %d", a, b)
	}
}

func TestPointsTierMissing(t *testing.T) {
	ct := weChaosTerminators(t)
	if pointsTierMissing(ct, 5, 1) || pointsTierMissing(ct, 5, 3) {
		t.Error("5 models should be covered at ordinals 1 and 3")
	}
	if !pointsTierMissing(ct, 4, 1) {
		t.Error("4 models is below the smallest tier")
	}
}

// Venatari Custodians: 3 models @160, or 4-6 models @320 (a GW range-priced tier
// with models_max=6). Every size in the range prices at 320.
func TestRangePricedTierVenatari(t *testing.T) {
	ds := EmbeddedDataset()
	ven, ok := ds.Units.GetInFaction("venatari-custodians", "adeptus-custodes")
	if !ok {
		t.Fatal("venatari-custodians not in dataset")
	}
	for _, c := range []struct{ models, want int }{{3, 160}, {4, 320}, {5, 320}, {6, 320}} {
		if got := baseUnitPoints(ven.Raw, c.models, 1); got != c.want {
			t.Errorf("baseUnitPoints(%d models) = %d, want %d", c.models, got, c.want)
		}
	}
	if !pointsTierMissing(ven.Raw, 2, 1) || !pointsTierMissing(ven.Raw, 7, 1) {
		t.Error("2 and 7 models fall outside every tier range")
	}
	if pointsTierMissing(ven.Raw, 4, 1) || pointsTierMissing(ven.Raw, 6, 1) {
		t.Error("4 and 6 models are covered by the 4-6 range tier")
	}
}

func TestBaseLoadoutLegalDefault(t *testing.T) {
	ds := EmbeddedDataset()
	bz, ok := ds.Units.Get("khorne-berzerkers")
	if !ok {
		t.Fatal("khorne-berzerkers not in dataset")
	}
	opts := bz.WargearOptions()
	lo := baseLoadout(bz.Raw, 10, opts, nil)
	if lo["bolt-pistol-khorne-berzerkers"] != 10 || lo["chainblade"] != 10 || len(lo) != 2 {
		t.Fatalf("base loadout should be the no-swap set, got %v", lo)
	}
	if v := validateLoadout(bz.Raw, 10, opts, lo, nil); len(v) != 0 {
		t.Fatalf("base loadout should validate clean, got %v", v)
	}
}
